feat: add C ndarray API and refactor for `stats/base/sdsnanmean`

0PrashantYadav0 opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: na
  - task: lint_repl_help status: na
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: passed
  - task: lint_c_examples status: passed
  - task: lint_c_benchmarks status: passed
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: na
  - task: lint_typescript_tests status: na
  - task: lint_license_headers status: passed ---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes. report:
  - task: run_javascript_examples status: na
  - task: run_c_examples status: na
  - task: run_cpp_examples status: na
  - task: run_javascript_readme_examples status: na
  - task: run_c_benchmarks status: na
  - task: run_cpp_benchmarks status: na
  - task: run_fortran_benchmarks status: na
  - task: run_javascript_benchmarks status: na
  - task: run_julia_benchmarks status: na
  - task: run_python_benchmarks status: na
  - task: run_r_benchmarks status: na
  - task: run_javascript_tests status: na ---

Resolves None

## Description

> What is the purpose of this pull request?

This pull request:

-   Adds C ndarray implementation for `stats/base/sdsnanmean`
-   Refactors `stats/base/sdsnanmean`
-   Adds C implementation section in README.md.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves None

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

0PrashantYadav0 committed:
feat: add C ndarray interface and refactor implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

0PrashantYadav0 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into ndarray-sdsnanmean

0PrashantYadav0 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into ndarray-sdsnanmean

0PrashantYadav0 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into ndarray-sdsnanmean

kgryte commented:
@0PrashantYadav0 Should this PR be closed out?

0PrashantYadav0 commented:
/stdlib-merge

0PrashantYadav0 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into ndarray-sdsnanmean

0PrashantYadav0 commented:
@kgryte This pr depend on [Pull-4727](https://github.com/stdlib-js/stdlib/pull/4727). So we have to wait for pr-4727 to merge before opening this.

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/sdsnanmean/README.md:
```diff
@@ -51,7 +51,7 @@ The [arithmetic mean][arithmetic-mean] is defined as
-#### sdsnanmean( N, x, stride )
+#### sdsnanmean( N, x, strideX )
 
 Computes the [arithmetic mean][arithmetic-mean] of a single-precision floating-point strided array `x`, ignoring `NaN` values and using extended accumulation.
```
```suggestion
Computes the [arithmetic mean][arithmetic-mean] of a single-precision floating-point strided array, ignoring `NaN` values and using extended accumulation.
```

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/sdsnanmean/README.md:
```diff
@@ -61,26 +61,24 @@ var Float32Array = require( '@stdlib/array/float32' );
 var x = new Float32Array( [ 1.0, -2.0, NaN, 2.0 ] );
 var N = x.length;
```
```suggestion
```

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/sdsnanmean/benchmark/c/benchmark.length.c:
```diff
@@ -94,15 +94,19 @@ static float rand_float( void ) {
+		if ( rand_double() < 0.2 ) {
+			x[ i ] = 0.0 / 0.0; // NaN
+		} else {
+			x[ i ] = ( rand_double() * 20000.0 ) - 10000.0;
```
```suggestion
			x[ i ] = ( rand_float() * 20000.0f ) - 10000.0f;
```

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/sdsnanmean/benchmark/c/benchmark.length.c:
```diff
@@ -94,15 +94,19 @@ static float rand_float( void ) {
 	for ( i = 0; i < len; i++ ) {
-		x[ i ] = ( rand_float()*20000.0f ) - 10000.0f;
+		if ( rand_double() < 0.2 ) {
+			x[ i ] = 0.0 / 0.0; // NaN
```
```suggestion
			x[ i ] = 0.0f / 0.0f; // NaN
```

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/sdsnanmean/benchmark/c/benchmark.length.c:
```diff
@@ -94,15 +94,19 @@ static float rand_float( void ) {
 
 	for ( i = 0; i < len; i++ ) {
-		x[ i ] = ( rand_float()*20000.0f ) - 10000.0f;
+		if ( rand_double() < 0.2 ) {
```
```suggestion
		if ( rand_float() < 0.2 ) {
```

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/sdsnanmean/benchmark/c/benchmark.length.c:
```diff
@@ -121,6 +125,44 @@ static double benchmark( int iterations, int len ) {
+	int i;
+
+	for ( i = 0; i < len; i++ ) {
+		if ( rand_double() < 0.2 ) {
```
```suggestion
		if ( rand_float() < 0.2 ) {
```

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/sdsnanmean/benchmark/c/benchmark.length.c:
```diff
@@ -121,6 +125,44 @@ static double benchmark( int iterations, int len ) {
+
+	for ( i = 0; i < len; i++ ) {
+		if ( rand_double() < 0.2 ) {
+			x[ i ] = 0.0 / 0.0; // NaN
```
```suggestion
			x[ i ] = 0.0f / 0.0f; // NaN
```

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/sdsnanmean/benchmark/c/benchmark.length.c:
```diff
@@ -121,6 +125,44 @@ static double benchmark( int iterations, int len ) {
+		if ( rand_double() < 0.2 ) {
+			x[ i ] = 0.0 / 0.0; // NaN
+		} else {
+			x[ i ] = ( rand_double() * 20000.0 ) - 10000.0;
```
```suggestion
			x[ i ] = ( rand_float() * 20000.0f ) - 10000.0f;
```

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/sdsnanmean/lib/index.js:
```diff
@@ -30,18 +30,16 @@
 * var x = new Float32Array( [ 1.0, -2.0, NaN, 2.0 ] );
 * var N = x.length;
```
```suggestion
```

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/sdsnanmean/README.md:
```diff
@@ -90,45 +88,39 @@ Note that indexing is relative to the first index. To introduce an offset, use [
-#### sdsnanmean.ndarray( N, x, stride, offset )
+#### sdsnanmean.ndarray( N, x, strideX, offsetX )
 
 Computes the [arithmetic mean][arithmetic-mean] of a single-precision floating-point strided array, ignoring `NaN` values and using extended accumulation and alternative indexing semantics.
```
```suggestion
Computes the [arithmetic mean][arithmetic-mean] of a single-precision floating-point strided array, ignoring `NaN` values and using extended accumulation and using alternative indexing semantics.
```

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/sdsnanmean/include/stdlib/stats/base/sdsnanmean.h:
```diff
@@ -31,7 +31,12 @@ extern "C" {
+float API_SUFFIX(stdlib_strided_sdsnanmean)( const CBLAS_INT N, const float *X, const CBLAS_INT strideX );
+
+/**
+* Computes the arithmetic mean of a single-precision floating-point strided array, ignoring `NaN` values and using extended accumulation and alternative indexing semantics.
```
```suggestion
* Computes the arithmetic mean of a single-precision floating-point strided array, ignoring `NaN` values and using extended accumulation and using alternative indexing semantics.
```

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/sdsnanmean/src/main.c:
```diff
@@ -0,0 +1,47 @@
+
+#include "stdlib/stats/base/sdsnanmean.h"
+#include "stdlib/blas/base/shared.h"
+#include "stdlib/strided/base/stride2offset.h"
```
```suggestion
#include "stdlib/strided/base/stride2offset.h"
#include "stdlib/stats/base/sdsnanmeanors.h"
```

gururaj1512 committed:
bench: update c benchmarks

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
fix: c implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
fix: examples

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
docs: update docs

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
chore: minor cleanup

0PrashantYadav0 commented:
@gururaj1512 Why we updated `rand_double` to `rand_float`.

gururaj1512 commented:
> @gururaj1512 Why we updated `rand_double` to `rand_float`.

We generate array of single precision random numbers in single precision packages.

0PrashantYadav0 commented:
@gururaj151 Sorry, I just got a little confused there.

gururaj1512 committed:
chore: minor cleanup

Signed-off-by: Gururaj Gurram <143020143+gururaj1512@users.noreply.github.com>

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into ndarray-sdsnanmean

kgryte commented:
@0PrashantYadav0 Mind addressing the CI failures in this PR?

0PrashantYadav0 commented:
Sure

0PrashantYadav0 committed:
chore: updated manifest file

0PrashantYadav0 committed:
chore: updated manifest file

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/sdsnanmean/lib/ndarray.js:
```suggestion
* Computes the arithmetic mean of a single-precision floating-point strided array, ignoring `NaN` values and using extended accumulation and alternative indexing semantics.
```

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/sdsnanmean/lib/ndarray.native.js:
```suggestion
* Computes the arithmetic mean of a single-precision floating-point strided array, ignoring `NaN` values and using extended accumulation and alternative indexing semantics.
```

gururaj1512 committed:
chore: minor cleanup

Signed-off-by: Gururaj Gurram <143020143+gururaj1512@users.noreply.github.com>

gururaj1512 committed:
chore: minor cleanup

Signed-off-by: Gururaj Gurram <143020143+gururaj1512@users.noreply.github.com>

gururaj1512 commented:
/stdlib merge

Planeshifter commented:
Looks like this PR has some merge conflicts that need to be resolved.
